File: agentctl/gateway_core/internal/wire/wire.go
// Package wire is the header-only fast path for the Go data plane: it lets the gateway route
// and tag Frames by touching the protobuf wire bytes directly, without deserializing the whole
// message (the oneof payload + nested submessages + the attributes map).
//
// It exists because of two facts about how the proxy actually uses a Frame (see proxy.go):
//
//   - routing reads exactly one field, session_id (field 1), and only on the first frame;
//   - the gateway writes exactly one field, attributes["canary_arm"] (the map at field 16), on
//     outbound frames.
//
// Every other frame is forwarded verbatim. So the hot path never needs the typed struct:
// SessionID scans field 1 and stops; SetCanaryArm appends one map entry to the tail. The frozen
// routing/identity header (fields 1-4) is never rewritten, so the wire contract the conformance
// suite pins is preserved byte-for-byte (proven in wire_test.go against the same fixtures).
package wire

import "google.golang.org/protobuf/encoding/protowire"

// RawFrame is an opaque, already-serialized Frame. It is the message type the zero-copy proxy
// path passes to grpc's RecvMsg/SendMsg so the transport hands over the wire bytes verbatim
// (paired with the passthrough Codec in the gateway package) instead of building a typed Frame.
type RawFrame struct{ B []byte }

// Frame field numbers (must match proto/agent_control.proto - the frozen envelope).
const (
	fieldSessionID  = 1  // string  (frozen header)
	fieldAttributes = 16 // map<string,string>
)

// CanaryArmKey is the attributes key the gateway stamps on outbound frames so the client/telemetry
// can see which canary arm served the stream. Must match agentctl/gateway/proxy.py + proxy.go.
const CanaryArmKey = "canary_arm"

// SessionID extracts Frame.session_id (field 1) by scanning the wire bytes, without building the
// Frame struct. Returns ("", false) if the field is absent or the bytes are malformed. session_id
// is the lowest field number, so on a well-formed frame it is the first tag and the scan returns
// after one field.
func SessionID(b []byte) (string, bool) {
	for len(b) > 0 {
		num, typ, n := protowire.ConsumeTag(b)
		if n < 0 {
			return "", false
		}
		b = b[n:]
		if num == fieldSessionID && typ == protowire.BytesType {
			v, n := protowire.ConsumeBytes(b)
			if n < 0 {
				return "", false
			}
			return string(v), true
		}
		// Not our field: skip its value and continue. Fields above session_id are rare on the
		// hot path (session_id is serialized first), so this loop usually runs once.
		n = protowire.ConsumeFieldValue(num, typ, b)
		if n < 0 {
			return "", false
		}
		b = b[n:]
	}
	return "", false
}

// SetCanaryArm returns b with attributes["canary_arm"]=arm injected, by appending a single
// map-entry occurrence of field 16 to the tail. This is wire-legal and decoder-safe: protobuf
// merges repeated occurrences of a map field, and on a duplicate key the last occurrence wins, so
// appending overrides any canary_arm the backend already set - identical to the typed
// out.Attributes["canary_arm"]=arm the slow path does. The frozen header (fields 1-4) is untouched.
//
// The input b is not mutated; the result is a fresh slice.
func SetCanaryArm(b []byte, arm string) []byte {
	entry := appendMapEntry(nil, CanaryArmKey, arm)
	out := make([]byte, 0, len(b)+len(entry)+2)
	out = append(out, b...)
	out = protowire.AppendTag(out, fieldAttributes, protowire.BytesType)
	out = protowire.AppendBytes(out, entry)
	return out
}

// appendMapEntry encodes a map<string,string> entry submessage {1:key, 2:value} onto dst.
func appendMapEntry(dst []byte, key, val string) []byte {
	dst = protowire.AppendTag(dst, 1, protowire.BytesType)
	dst = protowire.AppendString(dst, key)
	dst = protowire.AppendTag(dst, 2, protowire.BytesType)
	dst = protowire.AppendString(dst, val)
	return dst
}

File: agentctl/gateway/wire.py
"""Header-only fast path for the Python reference proxy - the mirror of the Go `wire` package.

The proxy only ever reads one field (session_id, field 1, for routing) and writes one
(attributes["canary_arm"], the map at field 16). So the hot path never needs the typed Frame:
``session_id`` scans field 1 and stops; ``set_canary_arm`` appends one map entry to the tail (map
fields merge, last key wins, so appending overrides - and the frozen header, fields 1-4, is never
rewritten). Pure stdlib protobuf wire encoding; parity with the typed path is pinned against the same
conformance fixtures the Go side uses.
"""
from __future__ import annotations

_FIELD_SESSION_ID = 1
_FIELD_ATTRIBUTES = 16
CANARY_ARM_KEY = "canary_arm"


def _read_varint(b: bytes, i: int) -> tuple[int, int]:
    """Decode a base-128 varint at offset i; return (value, next_offset)."""
    shift = result = 0
    while True:
        byte = b[i]
        result |= (byte & 0x7F) << shift
        i += 1
        if not byte & 0x80:
            return result, i
        shift += 7


def _append_varint(out: bytearray, v: int) -> None:
    while True:
        b = v & 0x7F
        v >>= 7
        if v:
            out.append(b | 0x80)
        else:
            out.append(b)
            return


def _skip_field(b: bytes, i: int, wiretype: int) -> int:
    if wiretype == 0:          # varint
        _, i = _read_varint(b, i)
        return i
    if wiretype == 2:          # length-delimited
        ln, i = _read_varint(b, i)
        return i + ln
    if wiretype == 1:          # 64-bit
        return i + 8
    if wiretype == 5:          # 32-bit
        return i + 4
    raise ValueError(f"unsupported wiretype {wiretype}")


def session_id(raw: bytes) -> str | None:
    """Scan Frame.session_id (field 1) from the wire bytes without building the message. Returns None
    if absent/malformed. session_id is the lowest field number, so normally the first tag read."""
    i, n = 0, len(raw)
    try:
        while i < n:
            tag, i = _read_varint(raw, i)
            field, wiretype = tag >> 3, tag & 7
            if field == _FIELD_SESSION_ID and wiretype == 2:
                ln, i = _read_varint(raw, i)
                return raw[i:i + ln].decode("utf-8")
            i = _skip_field(raw, i, wiretype)
    except (IndexError, ValueError):
        return None
    return None


def _append_len_delim(out: bytearray, field: int, payload: bytes) -> None:
    _append_varint(out, (field << 3) | 2)
    _append_varint(out, len(payload))
    out.extend(payload)


def _map_entry(key: str, val: str) -> bytes:
    e = bytearray()
    _append_len_delim(e, 1, key.encode("utf-8"))   # map key  = field 1
    _append_len_delim(e, 2, val.encode("utf-8"))   # map value = field 2
    return bytes(e)


def set_canary_arm(raw: bytes, arm: str) -> bytes:
    """Return raw with attributes["canary_arm"]=arm appended as one field-16 map entry. Decoder-safe
    (map occurrences merge, last key wins) and leaves the frozen header untouched."""
    out = bytearray(raw)
    _append_len_delim(out, _FIELD_ATTRIBUTES, _map_entry(CANARY_ARM_KEY, arm))
    return bytes(out)
